internal/jobworker/grpc/server.go

```
@@ -0,0 +1,230 @@
+	user, ok := jw.userSvc.User(ctx)
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, "unauthenticated")
+	}
+
+	if req.JobId == "" {
+		return nil, status.Error(codes.InvalidArgument, validator.Format("empty job ID"))
+	}
+
+	j, err := jw.fetchJob(ctx, user, req.JobId)
+	if err != nil {
+		return nil, err
+	}
+
+	if j.Status() != job.Running {
+		return nil, status.Error(codes.FailedPrecondition, "job is not running")
+	}
+
+	if err := jw.jobSvc.StopJob(ctx, j.ID); err != nil {
+		logger.Errorf("stop job; job: %s, error: %v", j.ID, err)
+		return nil, status.Error(codes.Internal, "error stopping job")
+	}
+
+	return &pb.StopResponse{}, nil
+}
+
+func (jw JobWorker) Status(ctx context.Context, req *pb.StatusRequest) (*pb.StatusResponse, error) {
+	user, ok := jw.userSvc.User(ctx)
+	if !ok {
+		return nil, status.Error(codes.Unauthenticated, "unauthenticated")
+	}
+
+	if req.JobId == "" {
+		return nil, status.Error(codes.InvalidArgument, validator.Format("empty job ID"))
+	}
+
+	j, err := jw.fetchJob(ctx, user, req.JobId)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.StatusResponse{
+		Status: &pb.StatusDetail{
+			Status:   toStatus(j.Status()),
+			ExitCode: int32(j.ExitCode()),
+		},
+	}, nil
+}
+
+func (jw JobWorker) Output(req *pb.OutputRequest, stream pb.JobWorkerService_OutputServer) error {
+	user, ok := jw.userSvc.User(stream.Context())
+	if !ok {
+		return status.Error(codes.Unauthenticated, "unauthenticated")
+	}
+
+	if req.JobId == "" {
+		return status.Error(codes.InvalidArgument, validator.Format("empty job ID"))
+	}
+
+	j, err := jw.fetchJob(stream.Context(), user, req.JobId)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(stream.Context())
+	defer cancel()
+
+	outputc := make(chan []byte, streamBuffer)
+	go func() {
+		if err := j.StreamOutput(ctx, outputc, chunkSize); err != nil {
+			logger.Errorf("streaming output from job; job: %s, error: %v", j.ID, err)
```

This error should go back to the caller, not get swallowed.